### PR TITLE
Set etcd url in cloud-config when external etcd is enabled

### PR DESCRIPTION
**Pull request by vignesh-goutham**
*Issue #, if available:*
https://github.com/aws/eks-anywhere-internal/issues/2116

*Description of changes:*
These changes will set a release-url for etcdadm to download the etcd tarball if the right version is not cached already in the OS during build time. This is especially useful, when user is creating a cluster with a really old OS template but with a new EKS-A version. This change will also only be in effect for clusters created/upgraded with 0.19.0 or above, to avoid node rollouts.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/common/common.go [hunk 1]
```diff
@@ -147,3 +149,24 @@ func GetCAPIBottlerocketSettingsConfig(config *v1alpha1.BottlerocketConfiguratio
 
 	return strings.Trim(string(marshaledConfig), "\n"), nil
 }
+
+func GetExternalEtcdReleaseUrl(clusterVersion string, versionBundle *cluster.VersionsBundle) (string, error) {
+	clusterVersionSemVer, err := semver.New(clusterVersion)
```
Comment (jiayiwang7) [inline pkg/providers/common/common.go:154]: i don't think we need this as it is validated here: https://github.com/aws/eks-anywhere/blob/9a356ed3aac6f284b5d81eaa4aaf5dbd59c9811d/pkg/api/v1alpha1/cluster.go#L884 (this function is called in both cli and controller i think)

File: pkg/clusterapi/etcd.go [hunk 2]
```diff
@@ -11,15 +11,20 @@ import (
...
 // SetUbuntuConfigInEtcdCluster sets up the etcd config in EtcdadmCluster.
-func SetUbuntuConfigInEtcdCluster(etcd *etcdv1.EtcdadmCluster, version string) {
+func SetUbuntuConfigInEtcdCluster(etcd *etcdv1.EtcdadmCluster, versionsBundle *cluster.VersionsBundle, clusterSpec *cluster.Spec) {
 	etcd.Spec.EtcdadmConfigSpec.Format = etcdbootstrapv1.Format("cloud-config")
 	etcd.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
-		Version:    version,
+		Version:    versionsBundle.KubeDistro.EtcdVersion,
 		InstallDir: "/usr/bin",
 	}
+	etcdUrl, _ := common.GetExternalEtcdReleaseUrl(string(*clusterSpec.Cluster.Spec.EksaVersion), versionsBundle)
```
Comment (jiayiwang7) [inline pkg/clusterapi/etcd.go:24]: shouldn't we handle the error?

File: pkg/providers/cloudstack/template.go [hunk 3]
```diff
@@ -224,6 +225,10 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec) (map[string]interface{}, erro
 		values["externalEtcd"] = true
 		values["externalEtcdReplicas"] = clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.Count
 		values["etcdSshUsername"] = etcdMachineSpec.Users[0].Name
+		etcdUrl, _ := common.GetExternalEtcdReleaseUrl(string(*clusterSpec.Cluster.Spec.EksaVersion), versionsBundle)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/template.go:228]: same for all the providers, shouldn't we handle the error?

File: pkg/api/v1alpha1/cluster_types.go [hunk 4]
```diff
@@ -77,8 +77,14 @@ type ClusterSpec struct {
...
 
-// DevBuildVersion is the version string for the dev build of EKS-A.
-const DevBuildVersion = "v0.0.0-dev"
+const (
+	// DevBuildVersion is the version string for the dev build of EKS-A.
+	DevBuildVersion = "v0.0.0-dev"
+
+	// EtcdUrlEksAVersion is the version from which the etcd url will be set
+	// for etcdadm to pull the etcd tarball if that binary isnt cached
+	EtcdUrlEksAVersion = "v0.19.0"
```
[the remark this answers is not in the document]
Comment (vignesh-goutham) [reply pkg/api/v1alpha1/cluster_types.go:86]: This is the version from which we will set the etcd url. So this is pretty much static and wont change. We will use this to compare with the cluster version to check if we need to set etcd url or not.

**Review by jiayiwang7 — APPROVED** [review]

**Pull request merged**